Firestore/core/src/firebase/firestore/local/local_serializer.cc

```diff
@@ -208,6 +211,123 @@ std::unique_ptr<model::NoDocument> LocalSerializer::DecodeNoDocument(
                                               version);
 }
 
+util::Status LocalSerializer::EncodeQueryData(
+    const QueryData& query_data, std::vector<uint8_t>* out_bytes) const {
+  HARD_ASSERT(QueryPurpose::kListen == query_data.purpose(),
+              "Only queries with purpose %s may be stored, got %s",
+              QueryPurpose::kListen, query_data.purpose());
+  Writer writer = Writer::Wrap(out_bytes);
+  EncodeQueryData(&writer, query_data);
+  return writer.status();
+}
+
+void LocalSerializer::EncodeQueryData(Writer* writer,
+                                      const QueryData& query_data) const {
+  if (!writer->status().ok()) return;
+
+  writer->WriteTag({PB_WT_VARINT, firestore_client_Target_target_id_tag});
+  writer->WriteInteger(query_data.target_id());
+
+  writer->WriteTag(
+      {PB_WT_STRING, firestore_client_Target_snapshot_version_tag});
+  writer->WriteNestedMessage([&](Writer* writer) {
+    rpc_serializer_.EncodeTimestamp(writer,
+                                    query_data.snapshot_version().timestamp());
+  });
+
+  writer->WriteTag({PB_WT_STRING, firestore_client_Target_resume_token_tag});
+  writer->WriteBytes(query_data.resume_token());
+
+  const Query& query = query_data.query();
+  if (query.IsDocumentQuery()) {
+    // TODO(rsgowman): Implement. Probably like this (once EncodeDocumentsTarget
+    // exists):
+    /*
+    writer->WriteTag({PB_WT_STRING, firestore_client_Target_documents_tag});
+    writer->WriteNestedMessage([&](Writer* writer) {
+      rpc_serializer_.EncodeDocumentsTarget(writer, query);
+    });
+    */
+    abort();
+  } else {
+    writer->WriteTag({PB_WT_STRING, firestore_client_Target_query_tag});
+    writer->WriteNestedMessage([&](Writer* writer) {
+      rpc_serializer_.EncodeQueryTarget(writer, query);
+    });
+  }
+}
+
+util::StatusOr<QueryData> LocalSerializer::DecodeQueryData(
+    const uint8_t* bytes, size_t length) const {
+  Reader reader = Reader::Wrap(bytes, length);
+  QueryData query_data = DecodeQueryData(&reader);
+  if (reader.status().ok()) {
+    return query_data;
+  } else {
+    return reader.status();
+  }
+}
+
+QueryData LocalSerializer::DecodeQueryData(Reader* reader) const {
+  if (!reader->status().ok()) return QueryData::Invalid();
```

Comment:
As a though experiment, it would be useful to think through if we can unify the public use of this decoder and the internal usage of this decoder (i.e. could we return `StatusOr<QueryData>` here?)

This would have the benefit that we wouldn't have to create "invalid" instances all over the place because reader.status() could be the negative case.

```c++
if (!reader->status().ok()) return reader->status();
```

Based on the advice below (about avoiding all this early return business) I think most of the tag reading in a loop becomes
```c++
Tag tag = reader->ReadTag();
if (!reader->status().ok()) break;
```
Then the final line could be
```c++
return reader->StatusOr(QueryData(...));
```